### PR TITLE
Add tutorial on circuit optimization, gate alignment, & spin echoes

### PR DESCRIPTION
**Pull request by rmlarose**
See tutorial intro for full description. Briefly, this tutorial explains how to use circuit optimizers and insert spin echoes to improve circuit fidelity as discussed in the [best practices guide](https://quantumai.google/cirq/google/best_practices#improving_circuit_fidelity), and shows a benchmark.

cc @tanujkhattar

### CODE REVIEW

File: docs/tutorials/google/spin_echoes.ipynb [hunk 1]
```diff
@@ -0,0 +1,1362 @@
+{
```
Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>Also "Conversion to Target Gateset" ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #52.        # Full circuit. Note: We are intentionally creating a bad circuit structure
</pre><p>Please add a brief explanation of what's the intentional bad part (<code>strategy = cirq.InsertStrategy.NEW</code>)  <span class="ql-cursor">﻿</span></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>The "first" and "second" register are confusing here because we pass <code>[(3, 2), (4, 2), (4, 1), (5, 1), (6, 1)]]</code>  as the qubit order to <code>create_benchmark_circuit</code>  and hence the coupling happens between <code>(3, 2)</code>  and <code>(4, 2)</code>  but in the printed diagram the ordering is sorted lexicographically and hence <code>(3, 2)</code>  is first, <code>(4, 1)</code>  is second and <code>(4, 2)</code>  is third, therefore coupling is b/w 1st and 3rd. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>For the sake of completeness, can we print out the gateset of the processor by querying the engine? There's also another implicit assumption in the cell below that <code>PhX</code>  gates are part of the target gateset. It would be nice to make it explicit what the target gateset actually is. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>But the constructed circuit above already contains only <code>sqrt(iSwaps)</code> and hence there's no need to convert it to the target gateset?</p><p> </p><p><code>MergeInteractionsToSqrtIswap</code> merges all consecutive (1 &amp; 2 qubit) interactions on two qubits into a unitary matrix and then decomposes this unitary using sqrt iswaps in an attempt to a) convert to the target gateset and b) reduce the circuit depth by reducing the number of operations.</p><p> </p><p>I suggest that we start with a circuit which contains 2q gates other than sqrt iswaps so we can actually demonstrate the basis conversion. We can write a small description explaining that the user is free to choose any one of the many available analytical decomposers in Cirq, each of which can have a specialized purpose. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p><code>MergeInteractionsToSqrtISwap</code>  removed the entangling gate b/w the ancilla and first qubit. I think <code>MergeInteractionsToSqrtISwap</code> might not be the right choice here because the ideal circuit unitary is   just <code>Identity</code> , so if we merged all operations (not just one/two qubit), and tried to decompose we'd just get an empty circuit as the unitary would be identity ?</p><p> </p><p>The fact that we are trying to run a circuit with an identity unitary to understand device specific properties make it hard to run meaningful optimizations here because every optimization will end up changing the meaning of what we want to achieve by running the circuit? </p><p> </p><p>I'm not sure if this is a clear example for demonstration purpose. I imagine other readers could also get confused and have similar questions. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>Again, let's hint that Cirq supports decomposing to many different target gatesets via analytical decomposition functions present in https://github.com/quantumlib/Cirq/tree/master/cirq-core/cirq/optimizers. As an example, users can use <code>two_qubit_matrix_to_sqrt_iswap_operations</code>  to compile to <code>sqrt(iSwap)</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>"Remove unnecessary gates" sounds very generic and overloaded in terms of what it means. Maybe we should be more specific like "Eject Single Qubit Operations" (or some other better name)? </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>Moment 2 has single qubit gate (<code>PhX(-0.75)^0.5</code>) and two qubit gate (<code>sqrt(iswap)</code>)</p><p> </p><p>Note that categories should be an <code>Iterable[Category]</code> and <code>Classifier = Callable[['cirq.Operation'], bool]</code> , so <code>categories = [lambda op : len(op.qubits) == 1, lambda op : len(op.qubits) == 2]</code> should do the intended stratification. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Added!</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Added comment.</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Done!</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Used a different two-qubit gate (<code>cirq_google.SYC</code>)  in this first example to show the effect of <code>MergeInteractionsToSwqrtIswap</code>.  Also added your nice description of what it does - thanks!</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Reworded and added href to <code>cirq.optimizers</code>  to hint at the larger support.</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Agreed. Changed to "Eject single-qubit operations" here and in the outline at the start of the tutorial.</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Fixed, thanks!</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Still removes the two-qubit gate between ancilla &amp; 1st qubit and may lead to some confusion as you note, but I think we settled offline on keeping this example circuit. </p>

**Comment by rmlarose** [comment]
Thanks @tanujkhattar! Ready for another look.

File: docs/tutorials/google/spin_echoes.ipynb [hunk 2]
```diff
@@ -0,0 +1,1347 @@
...
```
Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <blockquote>All measurements should be terminal and in a single moment.</blockquote><p>One liner explanation for this as well? Maybe because measurements are the slowest operations on the device? and we don't yet support non-terminal measurements?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>It would be nice if this example circuit had a few more operations on the ancilla instead of directly a measurement.</p><p> </p><p>Though this problem is mitigated in the benchmark examples because of a different circuit generated by <code>sqrt_iswap</code> instead of <code>SYC</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:1]: <p>I would imagine that <code>SynchronizeTerminalMeasurements</code> would be useful when we have multiple measurements as pushing them to the same last moment would be beneficial.</p><p> </p><p>In this case, since we only have a single measurement and it's the terminal operation on that qubit, pushing it at the end actually makes the circuit output worse and we could get a better fidelity by having it the first operation on the moment?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/4373/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

File: docs/tutorials/google/spin_echoes.ipynb (hunk 1 above)
Comment (tanujkhattar) [reply docs/tutorials/google/spin_echoes.ipynb:1]: Sounds good.

**Review by tanujkhattar — APPROVED** [review]
LGTM % nits. 

Tl;Dr - The fabricated example for demonstration has some areas which could potentially cause confusion to the readers, but as discussed offline, this is probably good enough for a tutorial especially because we expect mostly advanced users to be using this. 

I've added a couple comments for the sake of completeness, but due to lack of a better alternative circuit, I think we can go ahead and merge this. 

Thanks for pushing through the multiple review rounds, this is definitely a pretty hard tutorial to write!

File: docs/tutorials/google/spin_echoes.ipynb [hunk 3]
```diff
@@ -0,0 +1,1347 @@
...
+    "# Gates for spin echoes.\n",
+    "pi_pulses = [\n",
+    "    cirq.PhasedXPowGate(phase_exponent=p, exponent=1.0) for p in (-0.5, 0.0, 0.5, 1.0)\n",
+    "]\n",
+    "\n",
+    "\n",
+    "def create_benchmark_circuit(\n",
+    "    qubits: Sequence[cirq.GridQubit],\n",
+    "    cycles: int,\n",
+    "    twoq_gate: cirq.Gate = cirq.ISWAP ** 0.5,\n",
```
Comment (tanujkhattar) [inline docs/tutorials/google/spin_echoes.ipynb:226]: Please use `cirq.SQRT_ISWAP` instead.

File: docs/tutorials/google/spin_echoes.ipynb (hunk 2 above)
Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>I think it's actually because the benchmark example circuit has 7 qubits whereas this circuit has 5 qubits - sometimes the optimizer removes the two-qubit gate on the ancilla and sometimes it doesn't.</p><p> </p><p>I could make this circuit be on 7 qubits also so that the two-qubit gate is not removed. (I just tested this and it is indeed not removed.) This would cause the rest of the circuits in this section and descriptions of what the optimizers do on the circuits to change, however. (This is the downside of having such explanations on non-trivial circuits.) What do you think?</p>

Comment (tanujkhattar) [reply docs/tutorials/google/spin_echoes.ipynb:1]: If it's not too much pain and it's easy to add optimizer descriptions for the 7 qubit circuit, let's change it. Or else let's submit it. 

I don't have a strong preference. I leave the decision to you.

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Yeah aside from putting the measurement in the last moment, this circuit isn't the best to show off what <code>SynchronizeTerminalMeasurements</code>  does. My suggested fix is to have small circuits designed specifically to show what each optimizer does, e.g. a circuit with three staggered measurements fed into <code>SynchronizeTerminalMeasurements</code> that comes out with non-staggered measurements. I don't think doing this is in the spec for this tutorial though, so not sure on what action to take for this.</p><p> </p><p>(It is true that putting the ancilla measurement right at the start would give better fidelity, but that would be considered cheating for this benchmark.)</p>

Comment (tanujkhattar) [reply docs/tutorials/google/spin_echoes.ipynb:1]: Hmm, I guess we can leave it as it is here. It at least helps highlight what's available and advance users can evaluate based on context whether to use the optimization or not.

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Added!</p>

Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb:1]: <p>Quickly changed to this (well, 6 qubits). Cons: Fair bit of churn. Pros: Two-qubit gate is no longer removed and the example circuit looks less weird, and I can just do <code>circuit</code>  instead of <code>print(circuit.to_text_diagram(...)</code>  again, hopefully easier to review now :) </p><p> </p><p>Also note I changed the two-qubit gate from <code>cg.SYC</code>  to <code>cirq.ISWAP</code> so the inverse two-qubit gate in the circuit is clearer. </p>

File: docs/tutorials/google/spin_echoes.ipynb (hunk 3 above)
Comment (rmlarose) [reply docs/tutorials/google/spin_echoes.ipynb]: Done.

File: docs/tutorials/google/spin_echoes.ipynb (hunk 2 above)
Comment (tanujkhattar) [reply docs/tutorials/google/spin_echoes.ipynb:1]: Thanks, this looks great!

**Review by tanujkhattar — APPROVED** [review]
LGTM!
